wallet/txauthor/author.go

```diff
@@ -0,0 +1,203 @@
+
+	h "github.com/decred/dcrwallet/internal/helpers"
+	"github.com/decred/dcrwallet/wallet/internal/txsizes"
+)
+
+// InputSource provides transaction inputs referencing spendable outputs to
+// construct a transaction outputting some target amount.  If the target amount
+// can not be satisified, this can be signaled by returning a total amount less
+// than the target or by returning a more detailed error implementing
+// InputSourceError.
+type InputSource func(target dcrutil.Amount) (total dcrutil.Amount, inputs []*wire.TxIn, scripts [][]byte, err error)
+
+// InputSourceError describes the failure to provide enough input value from
+// unspent transaction outputs to meet a target amount.  A typed error is used
+// so input sources can provide their own implementations describing the reason
+// for the error, for example, due to spendable policies or locked coins rather
+// than the wallet not having enough available input value.
+type InputSourceError interface {
+	error
+	InputSourceError()
+}
+
+// Default implementation of InputSourceError.
+type insufficientFundsError struct{}
+
+func (insufficientFundsError) InputSourceError() {}
+func (insufficientFundsError) Error() string {
+	return "insufficient funds available to construct transaction"
+}
+
+// AuthoredTx holds the state of a newly-created transaction and the change
+// output (if one was added).
+type AuthoredTx struct {
+	Tx          *wire.MsgTx
+	PrevScripts [][]byte
+	TotalInput  dcrutil.Amount
+	ChangeIndex int // negative if no change
+}
+
+// ChangeSource provides P2PKH change output scripts for transaction creation.
+type ChangeSource func() ([]byte, error)
+
+// NewUnsignedTransaction creates an unsigned transaction paying to one or more
+// non-change outputs.  An appropriate transaction fee is included based on the
+// transaction size.
+//
+// Transaction inputs are chosen from repeated calls to fetchInputs with
+// increasing targets amounts.
+//
+// If any remaining output value can be returned to the wallet via a change
+// output without violating mempool dust rules, a P2PKH change output is
+// appended to the transaction outputs.  Since the change output may not be
+// necessary, fetchChange is called zero or one times to generate this script.
+// This function must return a P2PKH script or smaller, otherwise fee estimation
+// will be incorrect.
+//
+// If successful, the transaction, total input value spent, and all previous
+// output scripts are returned.  If the input source was unable to provide
+// enough input value to pay for every output any any necessary fees, an
+// InputSourceError is returned.
+//
+// BUGS: Fee estimation may be off when redeeming non-compressed P2PKH outputs.
+func NewUnsignedTransaction(outputs []*wire.TxOut, relayFeePerKb dcrutil.Amount,
+	fetchInputs InputSource, fetchChange ChangeSource) (*AuthoredTx, error) {
+
+	targetAmount := h.SumOutputValues(outputs)
+	estimatedSize := txsizes.EstimateSerializeSize(1, outputs, true)
```

Comment:
I would use a more generous estimate here, I think current wallet uses outputs*2 for the inputs instead of 1. Many wallets for Decred are dusty from PoS.